admin/modules/config/plugins.php

```diff
@@ -415,22 +544,52 @@
 
 	$install_uninstall = false;
 
-	if($mybb->input['action'] == "activate")
+	if($do_upgrade)
 	{
-		$message = $lang->success_plugin_activated;
+		$plugininfo = read_json_file(MYBB_ROOT."inc/plugins/{$codename}/plugin.json");
 
-		// Plugin is compatible with this version?
-		if($plugins->is_compatible($codename) == false)
+		// Check the plugin's compatibility with the current MyBB version
+		if($plugins->is_compatible($plugininfo['compatibility']) == false)
 		{
 			flash_message($lang->sprintf($lang->plugin_incompatible, $mybb->version), 'error');
-			admin_redirect("index.php?module=config-plugins");
+			admin_redirect('index.php?module=config-plugins');
+		}
+
+		// Try to archive plugin's themelet
+		require_once MYBB_ROOT.'inc/functions_themes.php';
+		if(!archive_themelet($codename, /*$is_plugin_themelet = */true, $err_msg))
+		{
+			flash_message($err_msg, 'error');
+			admin_redirect('index.php?module=config-plugins');
+		}
+
+		// Run any custom upgrade function as required
+		if($installed && function_exists("{$codename}_upgrade"))
+		{
+			call_user_func("{$codename}_upgrade");
+		}
+		$message = $lang->success_plugin_updated;
+	}
+
+	if($do_upgrade || $action == "activate")
+	{
+		if(!$do_upgrade)
+		{
+			$message = $lang->success_plugin_activated;
+		} else {
+			// Plugin is compatible with this version?
+			$plugininfo = read_json_file(MYBB_ROOT."inc/plugins/{$codename}/plugin.json");
```

Comment:
why do you read the json file twice in case of an upgrade? See line 549. It might make sense to add a new method `getPluginInfo(string $codename): ?array` which will perform the parsing of the json file and then return a cached version of this in case of subsequent calls.